benchmarking/commons/launch_remote_common.py

```diff
@@ -75,5 +78,40 @@ def sagemaker_estimator_args(
         type=int,
         default=0,
         help="Skip this number of initial jobs which would be launched",
-    )
+    ),
+    dict(
+        name="backoff_wait_time",
+        type=int,
+        default=600,
+        help=(
+            "If positive, ClientError[ResourceLimitExceeded] is caught, and we "
+            "try again after this many seconds. If 0 or negative, this is not "
+            "done"
+        ),
+    ),
 ]
+
+
+def fit_sagemaker_estimator(backoff_wait_time: int, estimator: EstimatorBase, **kwargs):
+    """
+    Runs ``estimator.fit(**kwargs)``. If ``backoff_wait_time > 0``, this is
+    wrapped into the ``backoff_boto_clienterror`` decorator, which makes sure
+    that if ``fit`` fails with ``ClientError`` of type "ResourceLimitExceeded",
+    we wait for ``backoff_wait_time`` seconds and try again (up to 100 times).
+    If ``backoff_wait_time <= 0``, the call of ``fit`` is not wrapped.
+
+    :param backoff_wait_time: See above.
+    :param estimator: SageMaker estimator to call ``fit`` for
+    :param kwargs: Arguments for ``estimator.fit``
+    """
+
+    @backoff_boto_clienterror(
+        errorname="ResourceLimitExceeded", length2sleep=backoff_wait_time
+    )
+    def fit_sagemaker_estimator_with_backoff(estimator: EstimatorBase, **kwargs):
+        estimator.fit(**kwargs)
+
+    if backoff_wait_time > 0:
+        fit_sagemaker_estimator_with_backoff(estimator, **kwargs)
+    else:
+        estimator.fit(**kwargs)
```

Comment:
I tihnk this is abusing decorators a little. I would try to bake the logic in directly here or otherwise use something along the lines of:

```
  if backoff_wait_time > 0:
      backoff_boto_clienterror(errorname="ResourceLimitExceeded", length2sleep=backoff_wait_time)(estimator.fit(**kwargs))
  else:
      estimator.fit(**kwargs)
```

Reply:
I don't know this stuff well

Reply:
You know now what we need, maybe you have a proposal

Reply:
We need something that triggers on `botocore.exceptions.ClientError`, and only if the error string contains "ResourceLimitExceeded". Any other exception, or `ClientError` with other message, should pass through

Reply:
I would suggest to bake the decorator logic directly into this function, for example as:
```
def fit_sagemaker_estimator(backoff_wait_time: int, estimator: EstimatorBase, ntimes_resource_wait: int = 100, length2sleep: int = 360, **kwargs):
    """
    Runs ``estimator.fit(**kwargs)``. If ``backoff_wait_time > 0``, this is
    wrapped into the ``backoff_boto_clienterror`` decorator, which makes sure
    that if ``fit`` fails with ``ClientError`` of type "ResourceLimitExceeded",
    we wait for ``backoff_wait_time`` seconds and try again (up to 100 times).
    If ``backoff_wait_time <= 0``, the call of ``fit`` is not wrapped.
    :param backoff_wait_time: See above.
    :param estimator: SageMaker estimator to call ``fit`` for
    :param kwargs: Arguments for ``estimator.fit``
    """

    # If backoff_wait_time is None, run standard fitting
    if backoff_wait_time == 0:
        estimator.fit(**kwargs)

    for idx in range(ntimes_resource_wait):
        try:
            return estimator.fit(**kwargs)
        except Exception as e:
            if not e == botocore.exceptions.ClientError:
                raise (e)
        logger.info(
            f"botocore.exceptions.ClientError[{errorname}] detected "
            f"when calling <{some_function.__name__}>. Waiting "
            f"{length2sleep / 60} minutes before retrying"
        )
        time.sleep(length2sleep)
```

Reply:
Alternatively, we can adapt the previously available decorator and then use it:

```
def backoff(error: Exception, ntimes_resource_wait: int = 100, length2sleep: float = 600):
    """
    Decorator that back offs for a fixed about of s after a given error is detected
    """
    def errorcatch(some_function):
        @functools.wraps(some_function)
        def wrapper(*args, **kwargs):
            for idx in range(ntimes_resource_wait):
                try:
                    return some_function(*args, **kwargs)
                except Exception as e:
                    if not e == error:
                        raise (e)

                logger.info(
                    f"{error} detected when calling <{some_function.__name__}>, waiting {length2sleep / 60} minutes before retring"
                )
                time.sleep(length2sleep)
                continue
        return wrapper

    return errorcatch
```

and the insides of the function becomes

```
    @backoff_boto_clienterror(
        error=botocore.exceptions.ClientError, length2sleep=backoff_wait_time
    )
    def fit_sagemaker_estimator_with_backoff(estimator: EstimatorBase, **kwargs):
        estimator.fit(**kwargs)

    if backoff_wait_time > 0:
        fit_sagemaker_estimator_with_backoff(estimator, **kwargs)
    else:
        estimator.fit(**kwargs)
```